### PR TITLE
Add interfaces and implementation to load and localize/filter search engines.

### PR DESCRIPTION
**Pull request by pocmo**

### CODE REVIEW

File: components/browser/search/src/main/java/mozilla/components/browser/search/provider/AssetsSearchEngineProvider.kt [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+ */
+class AssetsSearchEngineProvider(
+    private val localizationProvider: SearchLocalizationProvider,
+    private val filters: List<SearchEngineFilter> = emptyList()
+) : SearchEngineProvider {
+
+    /**
+     * Load search engines from this provider.
+     */
+    override suspend fun loadSearchEngines(context: Context): List<SearchEngine> {
```
Comment (sputnikci) [inline components/browser/search/src/main/java/mozilla/components/browser/search/provider/AssetsSearchEngineProvider.kt]: \[ktlint\] WARNING: \[modifier\-order\] Incorrect modifier order \(should be "suspend override"\) in column 5

Comment (sputnikci) [inline components/browser/search/src/main/java/mozilla/components/browser/search/provider/AssetsSearchEngineProvider.kt]: \[detekt\] INFO: \[style/ModifierOrder\] Modifier order should be: suspend override

File: components/browser/search/src/test/java/mozilla/components/browser/search/provider/localization/LocaleSearchLocalizationProviderTest.kt [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    fun `provider returns new data if default locale changes`() {
+        Locale.setDefault(Locale("de", "DE"))
+
+        val provider = LocaleSearchLocalizationProvider()
+
+        assertEquals("de", provider.language)
+        assertEquals("DE", provider.country)
+        assertNull(provider.region)
+    }
+}
```
Comment (sputnikci) [inline components/browser/search/src/test/java/mozilla/components/browser/search/provider/localization/LocaleSearchLocalizationProviderTest.kt]: \[detekt\] INFO: \[style/NewLineAtEndOfFile\] Checks whether files end with a line separator\.

File: components/browser/search/src/test/java/mozilla/components/browser/search/provider/localization/LocaleSearchLocalizationProviderTest.kt [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        // other tests that do not expect the locale to change.
+        cachedLocale = Locale.getDefault()
+    }
+
+    @After
+    fun restoreLocale() {
+        cachedLocale?.let { Locale.setDefault(it) }
+    }
+
+    @Test
```
Comment (sputnikci) [inline components/browser/search/src/test/java/mozilla/components/browser/search/provider/localization/LocaleSearchLocalizationProviderTest.kt]: \[detekt\] INFO: \[style/NamingConventionViolation\] All names in the codebase should be matching the naming convention of the codebase\.

File: components/browser/search/src/test/java/mozilla/components/browser/search/provider/localization/LocaleSearchLocalizationProviderTest.kt [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    fun `default in test is en-US`() {
+        val provider = LocaleSearchLocalizationProvider()
+
+        assertEquals("en", provider.language)
+        assertEquals("US", provider.country)
+        assertNull(provider.region)
+    }
+
+    @Test
```
Comment (sputnikci) [inline components/browser/search/src/test/java/mozilla/components/browser/search/provider/localization/LocaleSearchLocalizationProviderTest.kt]: \[detekt\] INFO: \[style/NamingConventionViolation\] All names in the codebase should be matching the naming convention of the codebase\.

**Comment by pocmo** [comment]
Oh, I didn't expect a coverage decrease. I'll clean up the patch and look at the coverage report in order to add some more tests.

File: components/browser/search/src/main/java/mozilla/components/browser/search/provider/AssetsSearchEngineProvider.kt (hunk 1 above)
Comment (pocmo) [reply components/browser/search/src/main/java/mozilla/components/browser/search/provider/AssetsSearchEngineProvider.kt]: He, that warning is super annoying. Android Studio shows a warning that it **should be** override suspend.

File: components/browser/search/src/test/java/mozilla/components/browser/search/provider/AssetsSearchEngineProviderTest.kt [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        }
+
+        val searchEngineProvider = AssetsSearchEngineProvider(localizationProvider)
+
+        val searchEngines = searchEngineProvider.loadSearchEngines(RuntimeEnvironment.application)
+
+        assertEquals(6, searchEngines.size)
+    }
+
+    @Test
```
Comment (sputnikci) [inline components/browser/search/src/test/java/mozilla/components/browser/search/provider/AssetsSearchEngineProviderTest.kt]: \[detekt\] INFO: \[style/NamingConventionViolation\] All names in the codebase should be matching the naming convention of the codebase\.

File: components/browser/search/src/test/java/mozilla/components/browser/search/provider/AssetsSearchEngineProviderTest.kt [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        }
+
+        val searchEngineProvider = AssetsSearchEngineProvider(localizationProvider, listOf(filter))
+
+        val searchEngines = searchEngineProvider.loadSearchEngines(RuntimeEnvironment.application)
+
+        assertEquals(4, searchEngines.size)
+    }
+
+    @Test
```
Comment (sputnikci) [inline components/browser/search/src/test/java/mozilla/components/browser/search/provider/AssetsSearchEngineProviderTest.kt]: \[detekt\] INFO: \[style/NamingConventionViolation\] All names in the codebase should be matching the naming convention of the codebase\.

File: components/browser/search/src/test/java/mozilla/components/browser/search/provider/AssetsSearchEngineProviderTest.kt [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+            val searchEngineProvider = AssetsSearchEngineProvider(localizationProviderWithRegion)
+            val searchEngines = searchEngineProvider.loadSearchEngines(RuntimeEnvironment.application)
+
+            assertEquals(7, searchEngines.size)
+            assertContainsSearchEngine("google-2018", searchEngines)
+            assertContainsNotSearchEngine("google", searchEngines)
+        }
+    }
+
+    @Test
```
Comment (sputnikci) [inline components/browser/search/src/test/java/mozilla/components/browser/search/provider/AssetsSearchEngineProviderTest.kt]: \[detekt\] INFO: \[style/NamingConventionViolation\] All names in the codebase should be matching the naming convention of the codebase\.

File: components/browser/search/src/test/java/mozilla/components/browser/search/provider/AssetsSearchEngineProviderTest.kt [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+            override val country: String = "US"
+            override val language = "en"
+            override val region: String? = null
+        }
+
+        val searchEngineProvider = AssetsSearchEngineProvider(localizationProvider)
+
+        val searchEngines = searchEngineProvider.loadSearchEngines(RuntimeEnvironment.application)
+
+        assertEquals(6, searchEngines.size)
```
Comment (sputnikci) [inline components/browser/search/src/test/java/mozilla/components/browser/search/provider/AssetsSearchEngineProviderTest.kt]: \[detekt\] INFO: \[style/MagicNumber\] Report magic numbers\. Magic number is a numeric literal that is not defined as a constant and hence it's unclear what the purpose of this number is\. It's better to declare such numbers as constants and give them a proper name\. By default, \-1, 0, 1, and 2 are not considered to be magic numbers\.

File: components/browser/search/src/test/java/mozilla/components/browser/search/provider/AssetsSearchEngineProviderTest.kt [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+            val localizationProviderWithoutRegion = object : SearchLocalizationProvider() {
+                override val country: String = "DE"
+                override val language = "de"
+                override val region: String? = null
+            }
+
+            val searchEngineProvider = AssetsSearchEngineProvider(localizationProviderWithoutRegion)
+            val searchEngines = searchEngineProvider.loadSearchEngines(RuntimeEnvironment.application)
+
+            assertEquals(7, searchEngines.size)
```
Comment (sputnikci) [inline components/browser/search/src/test/java/mozilla/components/browser/search/provider/AssetsSearchEngineProviderTest.kt]: \[detekt\] INFO: \[style/MagicNumber\] Report magic numbers\. Magic number is a numeric literal that is not defined as a constant and hence it's unclear what the purpose of this number is\. It's better to declare such numbers as constants and give them a proper name\. By default, \-1, 0, 1, and 2 are not considered to be magic numbers\.

File: components/browser/search/src/test/java/mozilla/components/browser/search/provider/AssetsSearchEngineProviderTest.kt [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+        val provider = object : SearchLocalizationProvider() {
+            override val country: String = "XX"
+            override val language = "xx"
+            override val region: String? = null
+        }
+
+        val searchEngineProvider = AssetsSearchEngineProvider(provider)
+        val searchEngines = searchEngineProvider.loadSearchEngines(RuntimeEnvironment.application)
+
+        assertEquals(6, searchEngines.size)
```
Comment (sputnikci) [inline components/browser/search/src/test/java/mozilla/components/browser/search/provider/AssetsSearchEngineProviderTest.kt]: \[detekt\] INFO: \[style/MagicNumber\] Report magic numbers\. Magic number is a numeric literal that is not defined as a constant and hence it's unclear what the purpose of this number is\. It's better to declare such numbers as constants and give them a proper name\. By default, \-1, 0, 1, and 2 are not considered to be magic numbers\.

File: components/browser/search/src/main/java/mozilla/components/browser/search/provider/filter/SearchEngineFilter.kt [hunk 11]
```diff
@@ -0,0 +1,11 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.browser.search.provider.filter
+
+import mozilla.components.browser.search.SearchEngine
+
```
Comment (csadilek) [inline components/browser/search/src/main/java/mozilla/components/browser/search/provider/filter/SearchEngineFilter.kt]: Let's add some JavaDocs to the public types and methods.

File: components/browser/search/src/test/java/mozilla/components/browser/search/provider/AssetsSearchEngineProviderTest.kt (hunk 8 above)
Comment (csadilek) [reply components/browser/search/src/test/java/mozilla/components/browser/search/provider/AssetsSearchEngineProviderTest.kt]: OK, I think this warning we're gonna have to disable. Some magic numbers are OK, Sputnik.

File: components/browser/search/src/main/java/mozilla/components/browser/search/provider/AssetsSearchEngineProvider.kt (hunk 1 above)
Comment (shyiko) [reply components/browser/search/src/main/java/mozilla/components/browser/search/provider/AssetsSearchEngineProvider.kt]: @pocmo this was fixed in [ktlint@0.16.0](https://github.com/shyiko/ktlint/blob/master/CHANGELOG.md#added-2).

**Pull request merged**